src/_guides/language/sound-dart.md

```diff
@@ -345,19 +345,21 @@ However, the app throws an exception at runtime because it is an error
 to assign a list of Dogs to a list of Cats.
 
 <aside class="alert alert-info" markdown="1">
-**Note:** As of release 1.21, only dartdevc implements these runtime checks,
+**Note:** As of release 1.24, only dartdevc implements these runtime checks,
 but support in other tools is coming.
 </aside>
 
 ## Type inference
 
 Does strong mode Dart mean that you _always_ have to specify a type?
 
-Actually, no. Strong mode Dart supports type inference. In some cases,
-the analyzer can infer types for fields, methods, local variables,
+No. Although _types_ are mandatory in strong mode,
+type _annotations_ are optional.
+The analyzer can infer types for fields, methods, local variables,
```

Comment:
Can the analyzer really (reliably) infer the types for fields?

Reply:
(asked again in email, just now)

Reply:
I don't believe that we infer the types for all fields, but will for final / const fields. @leafpetersen Am I remembering correctly?

Reply:
For final fields that seems reasonable (assuming that they are initialized via an initializing expression, not through the constructor).

Reply:
Yes, we infer types for fields.  We do this via two techniques.  

The first is override inference: if a field is declared with a type in a superclass, and then overridden without a declared type, then the override will inherit the type from the overridden field.

If override inference does not apply, then we will use the inferred type from the RHS to infer a type for the LHS.  If inferring the type of the RHS requires (recursively) the type of the thing that we are attempting to infer, then it is a static error.

Reply:
This was our original proposal for toplevel inference:

https://github.com/dart-lang/sdk/blob/8f2be46aac7705ab6a1024a795aaf977519a8be2/docs/language/informal/toplevel-inference.md

It did not allow inference for instance fields: only static fields and toplevel variables.  That proved too restrictive, and so we generalized it to what I described above.  This generalized version hasn't been specified yet.